src/Corax/Queries/BoostingMatch.Erasure.cs

```diff
@@ -0,0 +1,156 @@
+                }
+
+                [MethodImpl(MethodImplOptions.AggressiveInlining)]
+                static int FillFunc(ref BoostingMatch match, Span<long> matches)
+                {
+                    if (match._inner is BoostingMatch<TInner, TQueryScoreFunction> inner)
+                    {
+                        var result = inner.Fill(matches);
+                        match._inner = inner;
+                        return result;
+                    }
+                    return 0;
+                }
+
+                [MethodImpl(MethodImplOptions.AggressiveInlining)]
+                static int AndWithFunc(ref BoostingMatch match, Span<long> matches)
+                {
+                    if (match._inner is BoostingMatch<TInner, TQueryScoreFunction> inner)
+                    {
+                        var result = inner.AndWith(matches);
+                        match._inner = inner;
+                        return result;
+                    }
+                    return 0;
+                }
+
+                [MethodImpl(MethodImplOptions.AggressiveInlining)]
+                static void ScoreFunc(ref BoostingMatch match, Span<long> matches, Span<float> scores)
+                {
+                    if (match._inner is BoostingMatch<TInner, TQueryScoreFunction> inner)
+                    {
+                        inner.Score(matches, scores);
+                        match._inner = inner;
+                    }
+                }
+
+                FunctionTable = new FunctionTable(&FillFunc, &AndWithFunc, &ScoreFunc, &CountFunc);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static BoostingMatch Create<TInner, TQueryScoreFunction>(in BoostingMatch<TInner, TQueryScoreFunction> query)
+            where TInner : IQueryMatch
+            where TQueryScoreFunction : IQueryScoreFunction
+        {
+            return new BoostingMatch(query, StaticFunctionCache<TInner, TQueryScoreFunction>.FunctionTable);
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static BoostingMatch WithConstant<TInner>(in TInner inner, float value)
+            where TInner : IQueryMatch
+        {
+            static void ConstantScoreFunc(ref BoostingMatch<TInner, ConstantScoreFunction> match, Span<long> matches, Span<float> scores)
+            {
+                if (match._inner.IsBoosting)
+                {
+                    // We execute for the whole chain up to here in case there are more boosting clauses below.                                      
+                    match._inner.Score(matches, scores);
+
+                    // We need to multiply by a constant.                       
+                    float value = match._scorer.Value;
+                    for (int i = 0; i < scores.Length; i++)
+                    {
+                        // TODO: We will need to filter out the matches that do not belong to this branch. 
```

Comment:
What do you mean?

Reply:
You set the score only for those that match, not for all of them. It is a tradeoff, some more memory vs complexity.

Reply:
At this point, aren't we guaranteed that all the `matches` are in the final result?

Reply:
Not true in the case of OR and boosting only a single branch. 
`statement_a or statement_b.Boost(10)`
your output will be all matches from both... when you request the scoring, only those that have come from `statement_b` are going to have 10 assigned to them.